From b2ccddae6d23a6163c12dbdf9721d747bd95157d Mon Sep 17 00:00:00 2001
From: Alex Deucher <alexander.deucher@amd.com>
Date: Tue, 14 Mar 2017 17:57:35 -0400
Subject: [PATCH 1502/2094] drm/amdgpu/raven: power up/down VCN via the SMU
 (v2)

By default VCN is powered down like SDMA, power it up/down
on driver load/unload.

[Rui: Fix to add the parameter 0 to un-gate VCN] v2

Acked-by: Huang Rui <ray.huang@amd.com>
Reviewed-by: Alex Deucher <alexander.deucher@amd.com>
Reviewed-by: Hawking Zhang <Hawking.Zhang@amd.com>
Signed-off-by: Alex Deucher <alexander.deucher@amd.com>
Signed-off-by: Huang Rui <ray.huang@amd.com>
Signed-off-by: Alex Deucher <alexander.deucher@amd.com>
---
 drivers/gpu/drm/amd/powerplay/smumgr/rv_smumgr.c | 24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)

diff --git a/drivers/gpu/drm/amd/powerplay/smumgr/rv_smumgr.c b/drivers/gpu/drm/amd/powerplay/smumgr/rv_smumgr.c
index 6c931ba..b69dcb9 100644
--- a/drivers/gpu/drm/amd/powerplay/smumgr/rv_smumgr.c
+++ b/drivers/gpu/drm/amd/powerplay/smumgr/rv_smumgr.c
@@ -264,6 +264,27 @@ static int rv_smc_disable_sdma(struct pp_smumgr *smumgr)
 	return 0;
 }
 
+/* vcn is disabled by default in vbios, need to re-enable in driver */
+static int rv_smc_enable_vcn(struct pp_smumgr *smumgr)
+{
+	PP_ASSERT_WITH_CODE(!rv_send_msg_to_smc_with_parameter(smumgr,
+			PPSMC_MSG_PowerUpVcn, 0),
+			"Attempt to power up vcn Failed!",
+			return -EINVAL);
+
+	return 0;
+}
+
+static int rv_smc_disable_vcn(struct pp_smumgr *smumgr)
+{
+	PP_ASSERT_WITH_CODE(!rv_send_msg_to_smc_with_parameter(smumgr,
+			PPSMC_MSG_PowerDownVcn, 0),
+			"Attempt to power down vcn Failed!",
+			return -EINVAL);
+
+	return 0;
+}
+
 static int rv_smu_fini(struct pp_smumgr *smumgr)
 {
 	struct rv_smumgr *priv =
@@ -271,6 +292,7 @@ static int rv_smu_fini(struct pp_smumgr *smumgr)
 
 	if (priv) {
 		rv_smc_disable_sdma(smumgr);
+		rv_smc_disable_vcn(smumgr);
 		cgs_free_gpu_mem(smumgr->device,
 				priv->smu_tables.entry[WMTABLE].handle);
 		cgs_free_gpu_mem(smumgr->device,
@@ -288,6 +310,8 @@ static int rv_start_smu(struct pp_smumgr *smumgr)
 		return -EINVAL;
 	if (rv_smc_enable_sdma(smumgr))
 		return -EINVAL;
+	if (rv_smc_enable_vcn(smumgr))
+		return -EINVAL;
 
 	return 0;
 }
-- 
2.7.4

